### ISSUE TITLE
Rebuilding the kd-tree index is slow as the kd-tree grows; inserting and removing individual records from the kd-tree makes the algorithm run much slower

### ISSUE THREAD

**Issue by mjacar** [mjacar/pytorch-nec#3]
As the title states, building the index of the kd-tree gets to be quite slow as the kd-tree grows. The most obvious solution to this would seem to be to avoid rebuilding the kd-tree index by inserting and removing individual values. Unfortunately, based on my benchmarking, this makes calls to `nn_index` about 60% slower which causes the algorithm to run about 1000x slower in total. An investigation into the internals of FLANN or perhaps an investigation into alternatives to FLANN may be a wise idea.

**Comment by jlindsey15** [comment]
Hi!  I wanted to ask if you have any ideas for how to work on this issue, or other tips for speeding up the model

**Comment by mjacar** [comment]
Not sure, really. This investigation would be pretty open-ended. It would certainly start with completely grokking how kd-trees work which I will admit to putting zero effort into doing so far. From there, you can make a more informed decision about which approximate nearest neighbors library to use or even create your own data structure which might be more fit to purpose than anything off-the-shelf. I hope that's helpful. I'm not really sure what else to say.

**Comment by mjacar** [comment]
Just realized that you would probably be interested in this: https://github.com/rkern/line_profiler
That's what I used to profile this. With that you'll see that the kd-tree is far and away the hotspot of this application and optimizing any other portion will achieve basically nothing.

**Comment by jlindsey15** [comment]
Ok thanks.  For reference, do you have the implementation in which individual values are inserted/removed?  Is there any hope of getting that approach to work better?

**Comment by jlindsey15** [comment]
Also, do you have a sense of whether the speed would be improved significantly if there was no gradient descent adjustment of the stored keys/vals?

**Comment by jlindsey15** [comment]
Ok yeah thanks!  Would it make sense to try an implementing a version using addPoints and removePoint?  I think this might help even if one keeps backpropping through the keys/values, since you'd only be updating the nearest neighbors each time.  Does that sound like it makes sense? (Just trying to do a sanity check before I put the time in to implement this :p )

**Comment by jlindsey15** [comment]
Hm weird.  Have you tried/considered performing the updates to kd_tree only occasionally, e.g. every N steps or something?

**Comment by mjacar** [comment]
That would mean that the kd-tree index would be stale when we go to do a nearest neighbors lookup. I really don't think we would want that at all, even though it would certainly make things faster.

**Comment by jlindsey15** [comment]
The staleness would only be relevant for the determination of the identities of the nearest neighbors, right?  I.e. you might incorrectly assess which are the nearest neighbors because you are using an old index.  But the actual keys / values once they are determined would be correct since those are torch variables (I think?).  So maybe the staleness wouldn't be all that bad, if you do the update reasonably often, like every 10 steps or so?

**Comment by macardatto** [comment]
> The staleness would only be relevant for the determination of the identities of the nearest neighbors, right? I.e. you might incorrectly assess which are the nearest neighbors because you are using an old index. But the actual keys / values once they are determined would be correct since those are torch variables (I think?).

That's right.

> So maybe the staleness wouldn't be all that bad, if you do the update reasonably often, like every 10 steps or so?

I dunno. It seems kind of bad and would introduce behavior that would be difficult to reason about, but maybe it would work. 🤷‍♂️

**Comment by mjacar** [comment]
> The staleness would only be relevant for the determination of the identities of the nearest neighbors, right? I.e. you might incorrectly assess which are the nearest neighbors because you are using an old index. But the actual keys / values once they are determined would be correct since those are torch variables (I think?).

That's right.

> So maybe the staleness wouldn't be all that bad, if you do the update reasonably often, like every 10 steps or so?

I dunno. It seems kind of bad and would introduce behavior that would be difficult to reason about, but maybe it would work. 🤷‍♂️

**Comment by jlindsey15** [comment]
Ok.  Well I'll give it a try and see what happens.  Thanks for the guidance!
